python/feldera/sql_context.py

```diff
@@ -0,0 +1,365 @@
+
+        :param view_name: The name of the view to listen to.
+        """
+
+        queue = Queue(maxsize=1)
+
+        self.views_tx.append({view_name: queue})
+
+        handler = OutputHandler(self.client, self.pipeline_name, view_name, queue)
+        handler.start()
+
+        return handler
+
+    def from_delta_table(self, table_name: str, connector_name: str, config: dict):
+        """
+        Tells feldera to read the data from the specified delta table.
+
+        :param table_name: The name of the table.
+        :param connector_name: The unique name for this connector.
+        :param config: The configuration for the delta table.
+        """
+
+        # TODO: test this, can't test this right now, because of a serialization issue
+        # TODO: see: https://github.com/feldera/feldera/pull/1764
+
+        if config.get("uri") is None:
+            raise ValueError("uri is required in the config")
+
+        if config.get("mode") is None:
+            raise ValueError("mode is required in the config, valid modes: snapshot, follow, snapshot_and_follow")
+
+        if config.get("mode") not in ["snapshot", "follow", "snapshot_and_follow"]:
+            raise ValueError("mode must be one of snapshot, follow, snapshot_and_follow")
+
+        connector = Connector(name=connector_name,
+                              config={
+                                  "transport": {
+                                      "name": "delta_table_input",
+                                      "config": config,
+                                  }
+                              })
+
+        if table_name in self.input_connectors_buffer:
+            self.output_connectors_buffer[table_name].append(connector)
+        else:
+            self.output_connectors_buffer[table_name] = [connector]
+
+    def to_delta_table(self, view_name: str, connector_name: str, config: dict):
+        """
+        Tells feldera to write the data to the specified delta table.
+
+        :param view_name: The name of the view whose output is sent to delta table.
+        :param connector_name: The unique name for this connector.
+        :param config: The configuration for the delta table connector.
+        """
+
+        if config.get("uri") is None:
+            raise ValueError("uri is required in the config")
+
+        connector = Connector(name=connector_name,
+                              config={
+                                  "transport": {
+                                      "name": "delta_table_output",
+                                      "config": config,
+                                  }
+                              })
+
+        if view_name in self.output_connectors_buffer:
+            self.output_connectors_buffer[view_name].append(connector)
+        else:
+            self.output_connectors_buffer[view_name] = [connector]
+
+    def run_to_completion(self):
```

Comment:
This function leaves the pipeline in the running state (the `pipeline_complete` flag simply guarantees that all inputs have been processed, but the pipeline does not shut itself down automatically). It should probably call `shutdown` after reading all outputs. 

Even with this fix, it is still possible that the pipeline is already running when this function is called, in which case `start` is a noop even if the program and the connectors have changed. This function should either check the status and throw an exception if the pipeline is not in the shutdown state or force-shutdown it.